### DISCUSSION TITLE
Publish meal planner as iCal Calendar

### DISCUSSION THREAD

**Discussion by cashpw** [mealie-recipes/mealie#850] [Feature Request]
**Before submitting this feature request I have**
_COMPLETE AND FILL THESE OUT_
- [x] Reviewed the documentation and verified that my feature request isn't already a feature.
- [x] Reviewed the existing feature requests and my feature has not been requested.
- [x] Checked the [tasks taged](https://github.com/hay-kot/mealie/issues?q=is%3Aissue+is%3Aopen+label%3Atask+) issues and verified my feature is not covered

**Please Describe The Problem To Be Solved**

Users can only access their planned meals from within Mealie.

**(Optional): Suggest A Solution**

Publish the Meal Planner as an iCal for importing into Google Calendar, etc. See comments [1](https://github.com/hay-kot/mealie/discussions/318#discussioncomment-808634), [2](https://github.com/hay-kot/mealie/discussions/318#discussioncomment-1230537), [3](https://github.com/hay-kot/mealie/discussions/318#discussioncomment-1269057).

**Additional Information**

- [x] If the feature is accepted I'm willing to submit a PR to provide this feature

**Comment by hay-kot** [comment]
Heck yeah! 

FYI, you might have an easier time doing this with the next-branch given the rewrite of the data structure for meal-plans. I think we're about 1-2 months from a Release Candidate so up to you on where you want to do the work.

**Reply by cashpw** [reply]
I'd prefer to work with the new branch. I'll take a look and see what I can put together.  A few questions:

1. Is the next branch `mealie-next`?
2. Is there anything akin to a hook for database/api operations? My first thought was to trigger a "update the iCal file" function whenever the user updates the planner.

**Comment by cashpw** [comment]
Notes for myself:

- The [`icalendar`](https://icalendar.readthedocs.io/en/latest/index.html) extension seems applicable here.
- Don't bother limiting the date range at first -- show everything in the meal planner as this is what users would expect.

**Comment by bastantoine** [comment]
Hello @cashweaver! Have you started with this issue/plan on doing it in the near future, or can I tackle it?

**Comment by cashpw** [comment]
I'd welcome your tackling the issue. I've become incredibly busy recently.

On Thu, Apr 7, 2022 at 1:56 AM bastantoine ***@***.***> wrote:

> Hello @cashweaver <https://github.com/cashweaver>! Have you started with
> this issue/plan on doing it in the near future, or can I tackle it?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/hay-kot/mealie/discussions/850#discussioncomment-2521954>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANRKHUBAGEIG3J4HU2F4Y6LVD2PJ7ANCNFSM5I74BOQA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
five.sentenc.es, where possible.

Cheers,
Cash

**Comment by gbosky** [comment]
@bastantoine @cashpw Has there been any progress on this? I was starting to go down the mealie direction but one of the key features I am looking for is something that can integrate with Home Assistants calendar and/or iCal. Everything has been great with Mealie so far but I think this would be a great feature to make better use of my HA calendar.

**Comment by aureateflux** [comment]
I'd like this, but if possible: add direct caldav support (in the form of allowing us to enter the credentials for a specific caldav server and being able to publish directly to that).

Caldav servers usually support importing events via iCal, but being able to directly update would be good!

**Comment by btoconnor** [comment]
I was just about to submit my own Feature Request for this and found this using the search.  Having the ability to export the Meal Plan in some form of calendar would be amazing.  We're running into limitations where we want to start showing the calendar on some home dashboard screens, but the integrations don't exist for mealie - they only exist for regular calendars.

**Comment by boc-the-git** [comment]
This is a feature we'd accept into Mealie, as indicated by hay-kot when it was first raised.
If anyone wants to raise a PR, please do.
Any technical discussion/help needed, please jump in the Discord for the easiest support.

**Comment by hay-kot** [comment]
Here's some general notes on how this should get implemented 

1. Needs to support timezone selection per group, can default to timezone from machine.
2. Needs to support setting the event time (hours/minutes) for each meal plan type (Dinner, Lunch, etc...), so the user can decide when they show up on the calendar
3. Needs to support note types
4. Should have a link back to the recipe 
5. Should support filter meaplans by type to exclude notes, or breakfast from the calendar.

You could likely use this library https://github.com/ics-py/ics-py or an eq.

The endpoint would have to be unsecured so we'd also likely need to generate a random token to use in the URL to at least attempt to keep to information private.

**Comment by Cronocide** [comment]
Here's a little [Google Apps Script](https://gist.cronocide.net/Cronocide/mealie-sync) I run to sync my Mealie calendar to Google Calendar via their API, if anyone else needs such a thing while iCal support is in the backlog.

**Reply by Gullfrode** [reply]
I am trying to find your link, but get blocked

**Reply by Robubble** [reply]
Same for me. But it would be really helpful for to integrate this in a real calendar :)

**Comment by ubbitz** [comment]
Any progress on this?

**Comment by brent20** [comment]
Would love to see this! Thanks!

**Comment by weecoco2022** [comment]
Would love to see this
